wsmaster/che-core-api-agent/src/main/java/org/eclipse/che/api/agent/server/impl/AgentRegistryImpl.java

```diff
@@ -0,0 +1,130 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.agent.server.impl;
+
+import com.google.common.reflect.TypeToken;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import org.eclipse.che.api.agent.server.AgentRegistry;
+import org.eclipse.che.api.agent.server.AgentRegistryUrlProvider;
+import org.eclipse.che.api.agent.server.exception.AgentException;
+import org.eclipse.che.api.agent.server.exception.AgentNotFoundException;
+import org.eclipse.che.api.agent.shared.dto.AgentDto;
+import org.eclipse.che.api.agent.shared.model.Agent;
+import org.eclipse.che.api.agent.shared.model.AgentKey;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.api.core.rest.HttpJsonResponse;
+import org.eclipse.che.dto.server.DtoFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static org.eclipse.che.commons.lang.IoUtil.downloadFile;
+import static org.eclipse.che.commons.lang.IoUtil.readAndCloseQuietly;
+
+/**
+ * @author Anatoliy Bazko
+ */
+@Singleton
+public class AgentRegistryImpl implements AgentRegistry {
+    private final AgentRegistryUrlProvider urlProvider;
+    private final HttpJsonRequestFactory   requestFactory;
+
+    @Inject
+    public AgentRegistryImpl(AgentRegistryUrlProvider urlProvider,
+                             HttpJsonRequestFactory requestFactory) {
+        this.urlProvider = urlProvider;
+        this.requestFactory = requestFactory;
+    }
+
+    @Override
+    public Agent createAgent(String name, String version) throws AgentException {
+        URL url = getAgentUrl(name, Optional.ofNullable(version));
+        return doCreateRemoteAgent(url);
+    }
+
+    @Override
+    public Agent createAgent(AgentKey agentKey) throws AgentException {
+        URL url = getAgentUrl(agentKey.getName(), Optional.ofNullable(agentKey.getVersion()));
+        return doCreateRemoteAgent(url);
+    }
+
+    @Override
+    public Agent createAgent(String name) throws AgentException {
+        URL url = getAgentUrl(name, Optional.empty());
+        return doCreateRemoteAgent(url);
+    }
+
+    protected URL getAgentUrl(String name, Optional<String> version) throws AgentException {
```

Comment:
Disagree. If you have a wrapper it doesn't mean that you can't pass in a method.